File: app/task_store.py
from __future__ import annotations

import threading
import uuid
from datetime import datetime, timezone

from app.config import WORK_DIR_ROOT
from app.models import Task

_TASKS: dict[str, Task] = {}
_LOCK = threading.RLock()


def now_iso() -> str:
    return datetime.now(tz=timezone.utc).isoformat().replace('+00:00', 'Z')


def create_task(file_path: str, original_name: str) -> Task:
    task_id = str(uuid.uuid4())
    now = now_iso()
    task = Task(
        id=task_id,
        status='queued',
        filePath=file_path,
        workDir=str(WORK_DIR_ROOT / task_id),
        createdAt=now,
        updatedAt=now,
    )
    with _LOCK:
        _TASKS[task_id] = task
    log_task(task, f'Uploaded file: {original_name}')
    return task


def get_task(task_id: str) -> Task | None:
    with _LOCK:
        return _TASKS.get(task_id)


def list_tasks() -> list[Task]:
    with _LOCK:
        return sorted(_TASKS.values(), key=lambda t: t.createdAt, reverse=True)


def log_task(task: Task, message: str) -> None:
    stamp = now_iso()
    task.logs.append(f'[{stamp}] {message}')
    task.updatedAt = stamp


def set_task_error(task: Task, error: Exception | str, prefix: str) -> None:
    task.status = 'failed'
    raw = str(error)
    cleaned = ''.join(ch for ch in raw if ord(ch) >= 32 or ch in ('\n', '\r', '\t'))
    task.error = f'{prefix}: {cleaned}'
    log_task(task, task.error)

File: src/taskStore.ts
import fs from 'fs';
import path from 'path';
import { randomUUID } from 'node:crypto';
import { TASK_INDEX_PATH, WORK_DIR_ROOT } from './config';
import { Task } from './types';
import { normalizeSafeSegment } from './validators';

function readTasks(): Task[] {
  try {
    const raw = JSON.parse(fs.readFileSync(TASK_INDEX_PATH, 'utf8'));
    return Array.isArray(raw) ? (raw as Task[]) : [];
  } catch {
    return [];
  }
}

function writeTasks(tasks: Task[]): void {
  fs.writeFileSync(TASK_INDEX_PATH, `${JSON.stringify(tasks, null, 2)}\n`, 'utf8');
}

export function nowIso(): string {
  return new Date().toISOString();
}

function saveTask(nextTask: Task): Task {
  const tasks = readTasks();
  const index = tasks.findIndex(task => task.id === nextTask.id);
  if (index >= 0) {
    tasks[index] = nextTask;
  } else {
    tasks.push(nextTask);
  }
  writeTasks(tasks);
  return nextTask;
}

export function createTask(
  filePath: string,
  originalName: string,
  libraryItemId?: string | null,
  tenantId?: string,
  userId?: string | null,
): Task {
  const now = nowIso();
  const taskId = randomUUID();
  const safeTenantId = normalizeSafeSegment(tenantId || 'default');
  const task: Task = {
    id: taskId,
    status: 'queued',
    filePath,
    sourceName: originalName,
    workDir: path.join(WORK_DIR_ROOT, safeTenantId, taskId),
    createdAt: now,
    updatedAt: now,
    logs: [],
    libraryItemId: libraryItemId || null,
    tenantId: safeTenantId,
    userId: userId || null,
    errorCode: null,
    outputArtifactId: null,
    outputArtifactName: null,
  };
  logTask(task, `Uploaded file: ${originalName}`);
  return saveTask(task);
}

export function getTask(taskId: string): Task | undefined {
  return readTasks().find(task => task.id === taskId);
}

export function getTaskForTenant(taskId: string, tenantId?: string | null): Task | undefined {
  const safeTenantId = normalizeSafeSegment(tenantId || 'default');
  return readTasks().find(task => task.id === taskId && (task.tenantId || 'default') === safeTenantId);
}

export function listTasks(): Task[] {
  return readTasks().sort((a, b) => b.createdAt.localeCompare(a.createdAt));
}

export function updateTask(task: Task): Task {
  task.updatedAt = nowIso();
  return saveTask(task);
}

export function logTask(task: Task, message: string): Task {
  const entry = `[${nowIso()}] ${message}`;
  // also echo to console for easier debugging during development
  console.log(`task ${task.id}: ${message}`);
  task.logs.push(entry);
  return updateTask(task);
}

export function setTaskError(task: Task, error: unknown, prefix: string, code?: string): Task {
  task.status = 'failed';
  task.error = `${prefix}: ${String(error).replace(/[^\x09\x0A\x0D\x20-\uFFFF]/g, '')}`;
  task.errorCode = code || task.errorCode || 'TASK_FAILED';
  return logTask(task, task.error);
}
